### DISCUSSION TITLE
Preamble Detected Interrupt Help (again).

### DISCUSSION THREAD

**Discussion by jomar0** [jgromes/RadioLib#477] [Q&A]
Hey guys,
I'm having some trouble with the CAD DETECTED interrupt on my RFM95.
As far as I can tell, the library code (SX127x.cpp) keeps checking if DIO1 goes HIGH, while DIO0 is LOW.

The problem I'm having is that if I use startChannelScan() and digitalRead(DIO1) in the main loop, or use DIO1 as a wakeup source, or use a Logic Analyser, DIO1 never seems to go HIGH.

I know for a fact that it detects a LoRa preamble 100% of the time when I use scanChannel(). I also know that the pin monitoring code works, because a signal is detected when I short VCC and the GPIO pin.

I would really appreciate any help.

**Comment by jgromes** [comment]
Since `SX127x::scanChannel()`, is internally calling `startChannelScan()`, I don't see any reason it shouldn't work in your case. However, when going through the examples, I did find a missing `ICACHE_RAM_ATTR` attribute for one of the ISRs. It's fixed now, but doesn't explain the behavior you're seeing.

Also keep in mind that there's a timeout in CAD mode, so you have to monitor both DIO pins and periodically restart. You haven't posted your code, so I don't know if that could be the reason.

**Reply by jomar0** [reply]
Since I posted, I used the logic analyser and found the behaviour of the two pins,
it was very confusing because there are a number of 'blips', but I can clearly see that when CAD is detected, DIO1 goes high, and during CAD, DIO0 is low.  I programmed DIO3 and 4 to report CAD_DONE and CAD_detected, so I could confirm this behaviour.
Rather than use GPIO wakeup, I'm going to try coding the loop from scanChannel() into the ULP. I'm not sure if I can access the SPI device in the ULP, if I can't I'll just wake the device and restart the CAD.
Thanks for getting back to me.
